pkg/controller/clusterset/clusterset.go

```diff
@@ -0,0 +1,152 @@
+	Client      ctrlruntimeclient.Client
+	Scheme      *runtime.Scheme
+	ClusterCIDR string
+}
+
+// Add adds a new controller to the manager
+func Add(ctx context.Context, mgr manager.Manager, clusterCIDR string) error {
+	// initialize a new Reconciler
+	reconciler := ClusterSetReconciler{
+		Client:      mgr.GetClient(),
+		Scheme:      mgr.GetScheme(),
+		ClusterCIDR: clusterCIDR,
+	}
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&v1alpha1.ClusterSet{}).
+		WithOptions(controller.Options{
+			MaxConcurrentReconciles: maxConcurrentReconciles,
+		}).
+		Complete(&reconciler)
+}
+
+func (c *ClusterSetReconciler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	klog.Infof("%#v", req)
+	var clusterSet v1alpha1.ClusterSet
+	if err := c.Client.Get(ctx, types.NamespacedName{Name: req.Name, Namespace: req.Namespace}, &clusterSet); err != nil {
+		return reconcile.Result{}, fmt.Errorf("unable to get the clusterset: %w", err)
+	}
+
+	klog.Infof("got a clusterset: %v", clusterSet)
+	// create network policy
+	setNetworkPolicy, err := c.netpol(ctx, &clusterSet)
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("unable to make a networkpolicy for cluster set: %w", err)
+	}
+	if err := c.Client.Create(ctx, setNetworkPolicy); err != nil {
+		return reconcile.Result{}, fmt.Errorf("unable to create networkpolicy for clusterset: %w", err)
+	}
+	if clusterSet.Spec.MaxLimits != nil {
+		quota := v1.ResourceQuota{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "clusterset-quota",
+				Namespace: clusterSet.Namespace,
+				OwnerReferences: []metav1.OwnerReference{
+					{
+						UID:        clusterSet.UID,
+						Name:       clusterSet.Name,
+						APIVersion: clusterSet.APIVersion,
+						Kind:       clusterSet.Kind,
+					},
+				},
+			},
+		}
+		quota.Spec.Hard = clusterSet.Spec.MaxLimits
+		if err := c.Client.Create(ctx, &quota); err != nil {
+			return reconcile.Result{}, fmt.Errorf("unable to create resource quota from cluster set: %w", err)
+		}
+	}
+	return reconcile.Result{}, nil
+}
+
+func (c *ClusterSetReconciler) netpol(ctx context.Context, clusterSet *v1alpha1.ClusterSet) (*networkingv1.NetworkPolicy, error) {
+	var cidrList []string
+	if c.ClusterCIDR == "" {
+		var nodeList v1.NodeList
+		if err := c.Client.List(ctx, &nodeList); err != nil {
+			return nil, err
+		}
+		for _, node := range nodeList.Items {
+			cidrList = append(cidrList, node.Spec.PodCIDRs...)
+		}
+	} else {
+		cidrList = []string{c.ClusterCIDR}
+	}
+	return &networkingv1.NetworkPolicy{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      networkPolicyName,
+			Namespace: clusterSet.Namespace,
+		},
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "NetworkPolicy",
+			APIVersion: "networking.k8s.io/v1",
+		},
+		Spec: networkingv1.NetworkPolicySpec{
+			PolicyTypes: []networkingv1.PolicyType{
+				networkingv1.PolicyTypeIngress,
+				networkingv1.PolicyTypeEgress,
+			},
+			Ingress: []networkingv1.NetworkPolicyIngressRule{
+				{},
```

Comment:
This may be my ignorance speaking, but doesn't this block all ingress traffic, effectively preventing people from contacting the API server and using the cluster?

Reply:
the ingress is block all by default, if its a list tho with an empty item then it will allow all ingress or egress https://kubernetes.io/docs/concepts/services-networking/network-policies/#allow-all-ingress-traffic